lib/thin/async.rb

```diff
@@ -31,51 +31,58 @@ def schedule_dequeue
         end
     end
   end
-  
+
   # Response whos body is sent asynchronously.
   class AsyncResponse
     include Rack::Response::Helpers
-    
+
     Marker = [-1, {}, []].freeze
-    
+
     attr_reader :headers, :callback
     attr_accessor :status
-    
+
     def initialize(env, status=200, headers={})
       @callback = env['async.callback']
       @body = DeferrableBody.new
       @status = status
       @headers = headers
       @headers_sent = false
-      
+
       if block_given?
         yield self
         finish
       end
     end
-    
+
     def send_headers
       return if @headers_sent
       @callback.call [@status, @headers, @body]
       @headers_sent = true
     end
-    
+
     def write(body)
       send_headers
       @body.call(body.respond_to?(:each) ? body : [body])
     end
     alias :<< :write
-    
+
     # Tell Thin the response is complete and the connection can be closed.
     def done
+      return if done?
       send_headers
       EM.next_tick { @body.succeed }
+      @is_done = true
     end
-    
+
+    # Tells if the response has already been completed
+    def done?
+      !! @is_done
```

Comment:
`!!` is because it's initially `nil`. Concur on `is_done` vs. `done`. Bad habit of mine.